src/Extend/ServiceProvider.php

```diff
@@ -18,11 +18,14 @@ class ServiceProvider implements ExtenderInterface
 
     /**
      * Register a service provider.
+     * There are no backwards compatibility guarantees for service providers.
```

Comment:
I find our naming around service providers quite confusing.

It should be clear that the extender is fully part of the public API, it's what you might do with the service provider that might be outside of the public API. It's also my understanding that the `AbstractServiceProvider` class we provide with Flarum *is* part of the public API and won't break during the 1.x release cycle.

There are many legitimate use cases for service providers even for non-hacks, like registering singletons or custom container resolving for the extension's own class, or registering an external package into the container.

Maybe a warning in the form "Service providers are an advanced feature and might give access to Flarum internals that do not come with backward compatibility. Please read our documentation about service providers for recommendations" ?

Reply:
I like that wording!